sdk/lib/crt/misc/dbgrpt.cpp

```diff
@@ -0,0 +1,358 @@
+            OutputDebugStringA(filename ? filename : "<unknown file>");
+            OutputDebugStringA(", Line: ");
+            OutputDebugStringA(LineBuffer);
+            OutputDebugStringA("\n");
+
+            _CrtDbgBreak();
+
+            _InterlockedDecrement(&_CrtInAssert);
+            return -1;
+        }
+    }
+
+    if (filename)
+    {
+        _snprintf(szCompleteMessage, DBGRPT_MAX_BUFFER_SIZE, "%s(%d) : ", filename, linenumber);
+    }
+
+    if (format)
+    {
+        va_list arglist;
+        va_start(arglist, format);
+        int len = _vsnprintf(szFormatted, DBGRPT_MAX_BUFFER_SIZE - 2 - sizeof(DBGRPT_ASSERT_PREFIX_MESSAGE), format, arglist);
+        va_end(arglist);
+
+        if (len < 0)
+        {
+            strcpy(szFormatted, DBGRPT_STRING_TOO_LONG);
+        }
+
+        if (reportType == _CRT_ASSERT)
+            strcat(szCompleteMessage, DBGRPT_ASSERT_PREFIX_MESSAGE);
+        strcat(szCompleteMessage, szFormatted);
+    }
+    else if (reportType == _CRT_ASSERT)
+    {
+        strcat(szCompleteMessage, DBGRPT_ASSERT_PREFIX_NOMESSAGE);
+    }
+
+    if (reportType == _CRT_ASSERT)
+    {
+        if (_CrtModeOutputFormat[reportType] & _CRTDBG_MODE_FILE)
+            strcat(szCompleteMessage, "\r");
+        strcat(szCompleteMessage, "\n");
+    }
+
+    // FIXME: Handle user report hooks here
+
+
+    if (_CrtModeOutputFormat[reportType] & _CRTDBG_MODE_FILE)
+    {
+        OutputDebugStringA("ERROR: Please implement _CrtSetReportFile first\n");
+        _CrtDbgBreak();
+    }
+
+    if (_CrtModeOutputFormat[reportType] & _CRTDBG_MODE_DEBUG)
+    {
+        OutputDebugStringA(szCompleteMessage);
+    }
+
+    int nResult = FALSE;
+    if (_CrtModeOutputFormat[reportType] & _CRTDBG_MODE_WNDW)
+    {
+        nResult = _CrtDbgReportWindow(reportType, filename, linenumber, moduleName, szFormatted);
+
+        switch (nResult)
+        {
+        case IDRETRY:
+            nResult = TRUE;
+            break;
+        case IDIGNORE:
+        default:
+            nResult = FALSE;
+            break;
+        case IDABORT:
+            raise(SIGABRT);
+            _exit(3);
+            break;
+        }
+    }
+
+    if (reportType == _CRT_ASSERT)
+        _InterlockedDecrement(&_CrtInAssert);
+
+    return nResult;
+}
+
+EXTERN_C
+int __cdecl _CrtDbgReportW(int reportType, const wchar_t *filename, int linenumber, const wchar_t *moduleName, const wchar_t *format, ...)
+{
+    wchar_t szFormatted[DBGRPT_MAX_BUFFER_SIZE+1] = {0};           // The user provided message
+    wchar_t szCompleteMessage[(DBGRPT_MAX_BUFFER_SIZE+1)*2] = {0}; // The output for debug / file
+
+    if (reportType < 0 || reportType >= _CRT_ERRCNT)
+        return -1;
+
+    if (reportType == _CRT_ASSERT)
+    {
+        if (_InterlockedIncrement(&_CrtInAssert) > 0)
+        {
+            wchar_t LineBuffer[20] = {0};
+
+            _itow(linenumber, LineBuffer, 10);
+
+            OutputDebugStringW(L"Nested Assert from File: ");
```

Comment:
There's a balance here between consistent/readable code and efficiency, so you may already be doing the right thing...
... but just noting: OutputDebugStringW on a string literal is pretty wasteful since it gets converted to ansi anyway.

Reply:
I had not even considered efficiency for this,
since this is a corner case in debug logging code, so if this becomes a bottle neck, we probably have to fix something else.